File: test/language/choice_types/python/ChoiceCompatibilityCheckTest.py
import os
import zserio

import ChoiceTypes

from testutils import getApiDir

class ChoiceCompatibilityCheckTest(ChoiceTypes.TestCase):
    def testWriteVersion1ReadVersion1(self):
        choiceCompatibilityCheckVersion1 = self._createVersion1(self._createArrayVersion1())
        readChoiceCompatibilityCheckVersion1 = self._writeReadVersion1(choiceCompatibilityCheckVersion1)

        self.assertEqual(choiceCompatibilityCheckVersion1, readChoiceCompatibilityCheckVersion1)

    def testWriteVersion1ReadVersion2(self):
        choiceCompatibilityCheckVersion1 = self._createVersion1(self._createArrayVersion1())
        readChoiceCompatibilityCheckVersion2 = self._writeReadVersion2(choiceCompatibilityCheckVersion1)

        expectedArrayVersion2 = self._createArrayVersion2WithVersion1Fields()
        self.assertEqual(expectedArrayVersion2, readChoiceCompatibilityCheckVersion2.array)
        self.assertEqual(expectedArrayVersion2, readChoiceCompatibilityCheckVersion2.packed_array)

    def testWriteVersion2ReadVersion1(self):
        choiceCompatibilityCheckVersion2 = self._createVersion2(self._createArrayVersion2WithVersion1Fields())
        readChoiceCompatibilityCheckVersion1 = self._writeReadVersion1(choiceCompatibilityCheckVersion2)

        expectedArrayVersion1 = self._createArrayVersion1()
        self.assertEqual(expectedArrayVersion1, readChoiceCompatibilityCheckVersion1.array)
        self.assertEqual(expectedArrayVersion1, readChoiceCompatibilityCheckVersion1.packed_array)

    def testWriteVersion2ReadVersion2(self):
        choiceCompatibilityCheckVersion2 = self._createVersion2(self._createArrayVersion2())
        readChoiceCompatibilityCheckVersion2 = self._writeReadVersion2(choiceCompatibilityCheckVersion2)

        self.assertEqual(choiceCompatibilityCheckVersion2, readChoiceCompatibilityCheckVersion2)

    def testWriteVersion1ReadVersion1File(self):
        choiceCompatibilityCheckVersion1 = self._createVersion1(self._createArrayVersion1())
        readChoiceCompatibilityCheckVersion1 = self._writeReadVersion1File(choiceCompatibilityCheckVersion1,
                                                                           "version1_version1")

        self.assertEqual(choiceCompatibilityCheckVersion1, readChoiceCompatibilityCheckVersion1)

    def testWriteVersion1ReadVersion2File(self):
        choiceCompatibilityCheckVersion1 = self._createVersion1(self._createArrayVersion1())
        readChoiceCompatibilityCheckVersion2 = self._writeReadVersion2File(choiceCompatibilityCheckVersion1,
                                                                           "version1_version2")

        expectedArrayVersion2 = self._createArrayVersion2WithVersion1Fields()
        self.assertEqual(expectedArrayVersion2, readChoiceCompatibilityCheckVersion2.array)
        self.assertEqual(expectedArrayVersion2, readChoiceCompatibilityCheckVersion2.packed_array)

    def testWriteVersion2ReadVersion1File(self):
        choiceCompatibilityCheckVersion2 = self._createVersion2(self._createArrayVersion2WithVersion1Fields())
        readChoiceCompatibilityCheckVersion1 = self._writeReadVersion1File(choiceCompatibilityCheckVersion2,
                                                                           "version2_version1")

        expectedArrayVersion1 = self._createArrayVersion1()
        self.assertEqual(expectedArrayVersion1, readChoiceCompatibilityCheckVersion1.array)
        self.assertEqual(expectedArrayVersion1, readChoiceCompatibilityCheckVersion1.packed_array)

    def testWriteVersion2ReadVersion2File(self):
        choiceCompatibilityCheckVersion2 = self._createVersion2(self._createArrayVersion2())
        readChoiceCompatibilityCheckVersion2 = self._writeReadVersion2File(choiceCompatibilityCheckVersion2,
                                                                           "version2_version2")

        self.assertEqual(choiceCompatibilityCheckVersion2, readChoiceCompatibilityCheckVersion2)

    def _createVersion1(self, array):
        return self.api.ChoiceCompatibilityCheckVersion1(array, array)

    def _createVersion2(self, array):
        return self.api.ChoiceCompatibilityCheckVersion2(array, array)

    def _createArrayVersion1(self):
        return [
            self._createHolderVersion1(self.api.EnumVersion1.COORD_XY, 0),
            self._createHolderVersion1(self.api.EnumVersion1.TEXT, 1),
            self._createHolderVersion1(self.api.EnumVersion1.COORD_XY, 2),
            self._createHolderVersion1(self.api.EnumVersion1.TEXT, 3)
        ]

    def _createArrayVersion2WithVersion1Fields(self):
        return [
            self._createHolderVersion2(self.api.EnumVersion2.COORD_XY, 0),
            self._createHolderVersion2(self.api.EnumVersion2.TEXT, 1),
            self._createHolderVersion2(self.api.EnumVersion2.COORD_XY, 2),
            self._createHolderVersion2(self.api.EnumVersion2.TEXT, 3)
        ]

    def _createArrayVersion2(self):
        return self._createArrayVersion2WithVersion1Fields() + [
            self._createHolderVersion2(self.api.EnumVersion2.COORD_XYZ, 4),
            self._createHolderVersion2(self.api.EnumVersion2.COORD_XYZ, 5),
            self._createHolderVersion2(self.api.EnumVersion2.COORD_XYZ, 6)
        ]

    def _createHolderVersion1(self, selector, index):
        choice = self.api.ChoiceVersion1(selector)
        if selector == self.api.EnumVersion1.COORD_XY:
            choice.coord_xy = self.api.CoordXY(10 * index, 20 * index)
        else:
            choice.text = "text" + str(index)
        return self.api.HolderVersion1(selector, choice)

    def _createHolderVersion2(self, selector, index):
        choice = self.api.ChoiceVersion2(selector)
        if selector == self.api.EnumVersion2.COORD_XY:
            choice.coord_xy = self.api.CoordXY(10 * index, 20 * index)
        elif selector == self.api.EnumVersion2.TEXT:
            choice.text = "text" + str(index)
        else:
            choice.coord_xyz = self.api.CoordXYZ(10 * index, 20 * index, 1.1 * index)
        return self.api.HolderVersion2(selector, choice)

    def _writeReadVersion1(self, choiceCompatibilityCheck):
        writer = zserio.BitStreamWriter()
        choiceCompatibilityCheck.write(writer)

        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        return self.api.ChoiceCompatibilityCheckVersion1.from_reader(reader)

    def _writeReadVersion2(self, choiceCompatibilityCheck):
        writer = zserio.BitStreamWriter()
        choiceCompatibilityCheck.write(writer)

        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        return self.api.ChoiceCompatibilityCheckVersion2.from_reader(reader)

    def _writeReadVersion1File(self, choiceCompatibilityCheck, variant):
        filename = self.BLOB_NAME_BASE + variant + ".blob"
        zserio.serialize_to_file(choiceCompatibilityCheck, filename)
        return zserio.deserialize_from_file(self.api.ChoiceCompatibilityCheckVersion1, filename)

    def _writeReadVersion2File(self, choiceCompatibilityCheck, variant):
        filename = self.BLOB_NAME_BASE + variant + ".blob"
        zserio.serialize_to_file(choiceCompatibilityCheck, filename)
        return zserio.deserialize_from_file(self.api.ChoiceCompatibilityCheckVersion2, filename)

    BLOB_NAME_BASE = os.path.join(getApiDir(os.path.dirname(__file__)), "choice_compatibility_check_")

File: test/language/choice_types/cpp/ChoiceCompatibilityCheckTest.cpp
#include "gtest/gtest.h"

#include "choice_types/choice_compatibility_check/ChoiceCompatibilityCheckVersion1.h"
#include "choice_types/choice_compatibility_check/ChoiceCompatibilityCheckVersion2.h"

#include "zserio/SerializeUtil.h"

namespace choice_types
{
namespace choice_compatibility_check
{

using allocator_type = ChoiceCompatibilityCheckVersion1::allocator_type;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;

namespace
{
    template <typename HOLDER>
    struct VersionTraits
    {};

    template <>
    struct VersionTraits<HolderVersion1>
    {
        using Holder = HolderVersion1;
        using Choice = ChoiceVersion1;
        using Enum = EnumVersion1;
    };

    template <>
    struct VersionTraits<HolderVersion2>
    {
        using Holder = HolderVersion2;
        using Choice = ChoiceVersion2;
        using Enum = EnumVersion2;
    };
} // namespace

class ChoiceCompatibilityCheckTest : public ::testing::Test
{
protected:
    template <typename T, typename ARRAY>
    void fill(T& choiceCompatibilityCheck, const ARRAY& array)
    {
        choiceCompatibilityCheck.setArray(array);
        choiceCompatibilityCheck.setPackedArray(array);
    }

    template <typename T_READ, typename T_WRITE>
    T_READ writeRead(T_WRITE& choiceCompatibilityCheck)
    {
        zserio::BitStreamWriter writer(bitBuffer);
        choiceCompatibilityCheck.write(writer);

        zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
        return T_READ(reader);
    }

    template <typename T_READ, typename T_WRITE>
    T_READ writeReadFile(T_WRITE& choiceCompatibilityCheck, const char* variant)
    {
        const std::string fileName = BLOB_NAME_BASE + variant + ".blob";
        zserio::serializeToFile(choiceCompatibilityCheck, fileName);

        return zserio::deserializeFromFile<T_READ>(fileName);
    }

    template <typename HOLDER, typename ENUM>
    HOLDER createHolder(ENUM selector, uint32_t index)
    {
        using Choice = typename VersionTraits<HOLDER>::Choice;

        Choice choice;
        if (selector == ENUM::COORD_XY)
        {
            choice.initialize({ nullptr, 0, [](void*, size_t) { return ENUM::COORD_XY; } });
            choice.setCoordXY(CoordXY{10 * index, 20 * index});
        }
        else if (selector == ENUM::TEXT)
        {
            choice.initialize({ nullptr, 0, [](void*, size_t) { return ENUM::TEXT; } });
            choice.setText("text" + zserio::toString<allocator_type>(index));
        }

        return HOLDER(selector, std::move(choice));
    }

    template <typename HOLDER>
    vector_type<HOLDER> createArrayVersion1()
    {
        using Enum = typename VersionTraits<HOLDER>::Enum;

        return vector_type<HOLDER>{
            createHolder<HOLDER>(Enum::COORD_XY, 0),
            createHolder<HOLDER>(Enum::TEXT, 1),
            createHolder<HOLDER>(Enum::COORD_XY, 2),
            createHolder<HOLDER>(Enum::TEXT, 3)
        };
    }

    HolderVersion2 createHolderCoordXYZ(uint32_t index)
    {
        ChoiceVersion2 choice;
        choice.initialize({ nullptr, 0, [](void*, size_t) { return EnumVersion2::COORD_XYZ; } });
        choice.setCoordXYZ(CoordXYZ{10 * index, 20 * index, 1.1 * index});

        return HolderVersion2(EnumVersion2::COORD_XYZ, std::move(choice));
    }

    vector_type<HolderVersion2> createArrayVersion2()
    {
        auto array = createArrayVersion1<HolderVersion2>();
        array.push_back(createHolderCoordXYZ(4));
        array.push_back(createHolderCoordXYZ(5));
        array.push_back(createHolderCoordXYZ(6));
        return array;
    }



private:
    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);

    static const std::string BLOB_NAME_BASE;
};

const std::string ChoiceCompatibilityCheckTest::BLOB_NAME_BASE =
        "language/choice_types/choice_compatibility_check_";

TEST_F(ChoiceCompatibilityCheckTest, writeVersion1ReadVersion1)
{
    ChoiceCompatibilityCheckVersion1 choiceCompatibilityCheckVersion1;
    fill(choiceCompatibilityCheckVersion1, createArrayVersion1<HolderVersion1>());

    auto readChoiceCompatibilityCheckVersion1 = writeRead<ChoiceCompatibilityCheckVersion1>(
            choiceCompatibilityCheckVersion1);
    ASSERT_EQ(choiceCompatibilityCheckVersion1, readChoiceCompatibilityCheckVersion1);
}

TEST_F(ChoiceCompatibilityCheckTest, writeVersion1ReadVersion2)
{
    ChoiceCompatibilityCheckVersion1 choiceCompatibilityCheckVersion1;
    fill(choiceCompatibilityCheckVersion1, createArrayVersion1<HolderVersion1>());

    auto readChoiceCompatibilityCheckVersion2 = writeRead<ChoiceCompatibilityCheckVersion2>(
            choiceCompatibilityCheckVersion1);
    const auto expectedArrayVersion2 = createArrayVersion1<HolderVersion2>();
    ASSERT_EQ(expectedArrayVersion2, readChoiceCompatibilityCheckVersion2.getArray());
    ASSERT_EQ(expectedArrayVersion2, readChoiceCompatibilityCheckVersion2.getPackedArray());
}

TEST_F(ChoiceCompatibilityCheckTest, writeVersion2ReadVersion1)
{
    ChoiceCompatibilityCheckVersion2 choiceCompatibilityCheckVersion2;
    fill(choiceCompatibilityCheckVersion2, createArrayVersion1<HolderVersion2>());

    auto readChoiceCompatibilityCheckVersion1 = writeRead<ChoiceCompatibilityCheckVersion1>(
            choiceCompatibilityCheckVersion2);
    const auto expectedArrayVersion1 = createArrayVersion1<HolderVersion1>();
    ASSERT_EQ(expectedArrayVersion1, readChoiceCompatibilityCheckVersion1.getArray());
    ASSERT_EQ(expectedArrayVersion1, readChoiceCompatibilityCheckVersion1.getPackedArray());
}

TEST_F(ChoiceCompatibilityCheckTest, writeVersion2ReadVersion2)
{
    ChoiceCompatibilityCheckVersion2 choiceCompatibilityCheckVersion2;
    fill(choiceCompatibilityCheckVersion2, createArrayVersion2());

    auto readChoiceCompatibilityCheckVersion2 = writeRead<ChoiceCompatibilityCheckVersion2>(
            choiceCompatibilityCheckVersion2);
    ASSERT_EQ(choiceCompatibilityCheckVersion2, readChoiceCompatibilityCheckVersion2);
}

TEST_F(ChoiceCompatibilityCheckTest, writeVersion1ReadVersion1File)
{
    ChoiceCompatibilityCheckVersion1 choiceCompatibilityCheckVersion1;
    fill(choiceCompatibilityCheckVersion1, createArrayVersion1<HolderVersion1>());

    auto readChoiceCompatibilityCheckVersion1 = writeReadFile<ChoiceCompatibilityCheckVersion1>(
            choiceCompatibilityCheckVersion1, "version1_version1");
    ASSERT_EQ(choiceCompatibilityCheckVersion1, readChoiceCompatibilityCheckVersion1);
}

TEST_F(ChoiceCompatibilityCheckTest, writeVersion1ReadVersion2File)
{
    ChoiceCompatibilityCheckVersion1 choiceCompatibilityCheckVersion1;
    fill(choiceCompatibilityCheckVersion1, createArrayVersion1<HolderVersion1>());

    auto readChoiceCompatibilityCheckVersion2 = writeReadFile<ChoiceCompatibilityCheckVersion2>(
            choiceCompatibilityCheckVersion1, "version1_version2");
    const auto expectedArrayVersion2 = createArrayVersion1<HolderVersion2>();
    ASSERT_EQ(expectedArrayVersion2, readChoiceCompatibilityCheckVersion2.getArray());
    ASSERT_EQ(expectedArrayVersion2, readChoiceCompatibilityCheckVersion2.getPackedArray());
}

TEST_F(ChoiceCompatibilityCheckTest, writeVersion2ReadVersion1File)
{
    ChoiceCompatibilityCheckVersion2 choiceCompatibilityCheckVersion2;
    fill(choiceCompatibilityCheckVersion2, createArrayVersion1<HolderVersion2>());

    auto readChoiceCompatibilityCheckVersion1 = writeReadFile<ChoiceCompatibilityCheckVersion1>(
            choiceCompatibilityCheckVersion2, "version2_version1");
    const auto expectedArrayVersion1 = createArrayVersion1<HolderVersion1>();
    ASSERT_EQ(expectedArrayVersion1, readChoiceCompatibilityCheckVersion1.getArray());
    ASSERT_EQ(expectedArrayVersion1, readChoiceCompatibilityCheckVersion1.getPackedArray());
}

TEST_F(ChoiceCompatibilityCheckTest, writeVersion2ReadVersion2File)
{
    ChoiceCompatibilityCheckVersion2 choiceCompatibilityCheckVersion2;
    fill(choiceCompatibilityCheckVersion2, createArrayVersion2());

    auto readChoiceCompatibilityCheckVersion2 = writeReadFile<ChoiceCompatibilityCheckVersion2>(
            choiceCompatibilityCheckVersion2, "version2_version2");
    ASSERT_EQ(choiceCompatibilityCheckVersion2, readChoiceCompatibilityCheckVersion2);
}

} // namespace choice_compatibility_check
} // namespace choice_types
